src/azul/indexer/document.py

```diff
@@ -177,6 +193,10 @@ def __attrs_post_init__(self):
         concrete_type = type(self.bundle)
         assert concrete_type is BundleFQID, concrete_type
 
+    @property
+    def index_contents_type(self) -> IndexContentsType:
+        return Contributions(self.entity.entity_type)
```

Comment:
The smell here is that this is creating many instances that are all exactly the same. 

From what I understand, your motivation was to not have the string 'replica' twice:

> I disliked the redundancy of have to specify "replica" twice for both the discriminator and the entity type

The proposed solution is a big price to pay for what seems to me a minor redundancy.

Another smells is that there are now instanceof checks where a discriminator enum comparison would be a little more efficient.

In other words, I don't think you've nailed it yet.

You could intern these instances to reduce the number of redundant identical copies. It doesn't have to be perfect, just like its possible to create strings in Python that are equal but not the same. I caution though, that this might not be worth it. A simple discriminator enum might still be preferable.

Happy to discuss more in stand-up.